runtime-sdk/src/crypto/signature/sr25519.rs

```diff
@@ -70,3 +92,49 @@ impl From<&'static str> for PublicKey {
         PublicKey::from_bytes(&BASE64_STANDARD.decode(s).unwrap()).unwrap()
     }
 }
+
+/// A memory-backed signer for Sr25519.
+pub struct MemorySigner {
+    keypair: schnorrkel::Keypair,
+}
+
+impl Signer for MemorySigner {
+    fn random(rng: &mut (impl RngCore + CryptoRng)) -> Result<Self, Error> {
+        Ok(Self {
+            keypair: schnorrkel::Keypair::generate_with(rng),
+        })
+    }
+
+    fn new_from_seed(seed: &[u8]) -> Result<Self, Error> {
+        let seed = sha2::Sha512_256::digest(seed);
+        let mut rng = rand_chacha::ChaChaRng::from_seed(seed.into());
```

Comment:
Where is this defined or taken from?

Reply:
Just an ad-hoc deterministic way of generating a key. There is no spec for generating something from a seed AFAIK and we need this in tests. But we could also move some of this to `MemorySigner::new_test` and just forward to `from_bytes` here?

Reply:
Actually, we need this (or similar) as otherwise it is harder to quickly generate valid keys. @CedarMist any thoughts on this?

Reply:
Once we define this function, we cannot change it as if someone started using this `Signer`, new generation won't be compatible. So the best way would be to return signer in tests and error otherwise.

Reply:
The thing is that `new_from_seed` should accept some sort of (random) seed without structure. But if we make it forward to `from_bytes` here, it will require a specific structure.

Reply:
With the `generate` precompile, I look at it as a `Multiply base point by scalar` precompile.

Any modification to the scalar before point multiplication will cause problems.

Reply:
Can you provide this in terms of [`schnorrkel`](https://docs.rs/schnorrkel/0.11.4/schnorrkel/)?

Reply:
E.g. this could use `MiniSecretKey` and expansion, similar to Ed25519? Then the question is which kind of expansion, uniform or ed25519-like?

Reply:
This does not match Polkadot.

See: https://github.com/paulmillr/micro-sr25519/blob/017d85616c749b3714c3fba3c80b6c458729cc58/index.ts#L238

```typescript
export function secretFromSeed(seed: Uint8Array): Uint8Array {
  abytes('seed', seed, 32);
  const r = sha512(seed);
  // NOTE: different from ed25519
  r[0] &= 248;
  r[31] &= 63;
  r[31] |= 64;
  // this will strip upper 3 bits and lower 3 bits
  const key = encodeScalar(decodeScalar(r.subarray(0, 32)));
  const nonce = r.subarray(32, 64);
  return concatBytes(key, nonce);
}
```

See the test vectors at: https://github.com/polkadot-js/wasm/blob/10010830094e7d033bd11b16c5e3bc01a7045309/packages/wasm-crypto/src/rs/sr25519.rs#L176

PaulMillr's sr25519 library matches the Polkadot wasm-crypto package.

Reply:
Changed derivation to use Ed25519-based expansion as implemented by `schnorrkel` (and also used by the linked WASM implementation).